### PR TITLE
Expose internal variables

### PR DESCRIPTION
**Pull request by devloper-eth**
Great hardhat plugin! It makes solidity unit testing within waffle/truffle much easier. ❤️

I updated the code to also expose internal variables. I never used the solidity-last lib before, so I hope I got it right. Would love to discuss.

### CODE REVIEW

File: src/core.ts [hunk 1]
```diff
@@ -186,7 +202,10 @@ function getFunctionArguments(fnDef: FunctionDefinition): Argument[] {
 }
 
 function getType(varDecl: VariableDeclaration, location: StorageLocation = varDecl.storageLocation): string {
-  const { typeString, typeIdentifier } = varDecl.typeDescriptions;
```
Comment (devloper-eth) [inline src/core.ts:189]: `varDecl.typeDescriptions`  didn't include the `_ptr` suffix, but `varDecl.typeName.typeDescriptions` did.

File: src/core.ts [hunk 2]
```diff
@@ -208,6 +227,24 @@ function mapContracts(solcOutput: SolcOutput): ContractMap {
...
+  const parents = contract.linearizedBaseContracts.map(id => mustGet(contractMap, id));
+
+  const overriden = new Set<number>();
+  const res = [];
+
+  for (const parent of parents) {
+    for (const v of findAll('VariableDeclaration', parent)) {
+      if (v.stateVariable && v.visibility === 'internal' && !overriden.has(v.id)) {
+        res.push(v);
+        overriden.add(v.id);
```
Comment (devloper-eth) [inline src/core.ts:240]: Does this make sense?

**Comment by devloper-eth** [comment]
PR would need cleanup and refactoring. But curious to hear some thoughts and feedback.

File: src/core.ts (hunk 2 above)
Comment (frangio) [reply src/core.ts:245]: This can be removed. It's used for internal function but it doesn't make sense for variables.

File: src/core.ts [hunk 3]
```diff
@@ -76,10 +76,27 @@ function getExposedContent(ast: SourceUnit, inputPath: string, contractMap: Cont
...
+                `x${v.name}(${getGetterTypes(v).map(a => `${a.type} ${a.name}`).join(', ')})`,
+                'external',
+                'view',
+                'returns',
+              ];
+              header.push(`(${getGetterValues(v).map(a => `${a.type}`)})`);
+              header.push('{');
+              return [
+                header.join(' '), [
+                  `return ${v.name}` + (getGetterTypes(v).length > 0 ? `${getGetterTypes(v).map(a => `[${a.name}]`).join('')}` : '') + ';',
```
Comment (frangio) [inline src/core.ts:96]: You don't really need the conditional here. If length is 0, the `map(...).join('')` will by itself result in an empty string.

File: src/core.ts [hunk 4]
```diff
@@ -208,6 +232,63 @@ function mapContracts(solcOutput: SolcOutput): ContractMap {
...
+}
+
+function getGetterTypes(v: VariableDeclaration): Argument[] {
+  if (!v.typeName) {
+    throw new Error('missing typenName');
+  }
+  return _getGetterTypes(v.typeName, []);
+}
+
+function _getGetterTypes(v: TypeName, types: Argument[]): Argument[] {
```
Comment (frangio) [inline src/core.ts:260]: Can you inline this function into `getGetterTypes`? The same for `_getMappingValues` with `getGetterValues`.

File: src/core.ts (hunk 1 above)
Comment (frangio) [reply src/core.ts:189]: What is the relevance of the `_ptr` suffix? I don't really understand.

**Review by frangio — COMMENTED** [review]
Hey, thanks! Sorry I just now reviewed this. I think it's a good idea. With a little cleanup I'd love to merge it.

File: src/core.ts (hunk 1 above)
Comment (frangio) [reply src/core.ts:189]: What is the relevance of the `_ptr` suffix? I don't really understand.

File: src/core.ts (hunk 2 above)
Comment (frangio) [reply src/core.ts:245]: This can be removed. It's used for internal function but it doesn't make sense for variables.

**Comment by frangio** [comment]
Thanks for the initiative on this! Closing in favor of #4.

**Comment by frangio** [comment]
Published in the latest version!

**Pull request closed without merging**